config-service-impl/src/main/java/org/hypertrace/config/service/store/ConfigStore.java

```diff
@@ -23,34 +26,42 @@
   /**
    * Write the config value associated with the specified config resource to the store.
    *
-   * @param configResource
+   * @param configResourceContext
    * @param userId
    * @param config
-   * @return the context specific config written to the store
+   * @return the config written to the store
    */
-  InternalContextSpecificConfig writeConfig(
-      ConfigResource configResource, String userId, Value config) throws IOException;
+  UpsertedConfig writeConfig(
+      ConfigResourceContext configResourceContext, String userId, Value config) throws IOException;
 
   /**
    * Get the config with the latest version for the specified resource.
    *
-   * @param configResource
+   * @param configResourceContext
    * @return
    */
-  ContextSpecificConfig getConfig(ConfigResource configResource) throws IOException;
+  ContextSpecificConfig getConfig(ConfigResourceContext configResourceContext) throws IOException;
 
   /**
    * Get all the configs with the latest version(along with the context to which it applies) for the
    * specified parameters, sorted in the descending order of their creation time.
    *
-   * @param resourceName
-   * @param resourceNamespace
-   * @param tenantId
+   * @param configResource
    * @return
    * @throws IOException
    */
-  List<ContextSpecificConfig> getAllConfigs(
-      String resourceName, String resourceNamespace, String tenantId) throws IOException;
+  List<ContextSpecificConfig> getAllConfigs(ConfigResource configResource) throws IOException;
+
+  /**
+   * Write each of the provided config value associated with the specified config resource to the
+   * store.
+   *
+   * @param resourceContextValueMap
+   * @param userId
+   * @return the upserted configs
+   */
+  List<UpsertedConfig> writeAllConfigs(
```

Comment:
what is the error handling behavior ?

Reply:
Currently, it's identical to calling N write operations sequentially. If a later one fails, it aborts at that point leaving the original (this is the same behavior in the current abstraction one layer up in ObjectStore). Ideally, with a doc store API change we can make this transactional with all or nothing semantics (this came up with Entity Service in conversations with @saxenakshitiz too, but I believe as of today doc store itself doesn't support this - another reason I paused this change where I did).